### PR TITLE
iox-#27 Run `ServiceDiscovery` tests for `popo::UntypedServer`

### LINKED ISSUE #27: Request/Response communication with iceoryx
## Brief feature description

Besides publish/subscribe also request/response communication shall be supported

## Detailed information

So far iceoryx only supports publish/subscribe. For frameworks like ROS2 also a request/response communication shall be supported (services in ROS2).  The building blocks in posh can be re-used here, so start with client and server ports and then do the layers above

We will have the same layered architecture as for publishers and subscribers. With `ChunkSender` and `ChunkReceiver` common building blocks are reused but a `Client` and a `Server` will use one of each for the bidirectional data transfer with requests and responses. 
This image is a simplified view that shows the class hierarchy and the public method names for the typed C++ API. 

component

1. It will also have a zero-copy API, i.e. we can loan requests and take responses on the `Client` side and take requests and loan responses on the `Server` side.   
1. Like a `Publisher`, a `Server` can `offer()` or `stopOffere()`
2. Similar to a `Subscriber` `subscribe()` and `unsubscribe()`, the `Client` can `connect()`, `disconnect()` and getConnectionState()
3. You can poll the `Server` for requests and the `Client` for responses or use a `Waitset` to wait for requests and responses 

- [x] 1. Design and interface implementation for ports

- [x] 2. Implementation and unit tests for ports

- [x] 3. Integration test for ports
   - [x] including QueueFullPolicy tests with a real Client- and ServerPort

- [x] 4. C API 
   - [x] client
   - [x] request/response header
   - [x] server
   - [x] examples
   - [ ] (optional) add `iceoryx_posh_types.hpp` RPC related config stuff to `config.h`.
   - [ ] (optional) add missing functionality to chunk header

- [x] 5. C++ API
   - [x] typed C++ API 
   - [x] tests for typed C++ API
   - [x] untyped C++ API
   - [x] tests for untyped C++ API
   - [x] send returns an expected instead of void

- [x] 6. Integration in runtime and RouDi
   - [x] Make `ServerPort`'s findable via `ServiceDiscovery` 

- [x] 7. Common port queue policies
    - rename `SubscriberTooSlowPolicy` -> `ConsumerTooSlowPolicy`
        - alias for ConsumerTooSlowPolicy for easy migration to new API
        - rename `WAIT_FOR_SUBSCRIBER` -> `WAIT_FOR_CONSUMER`
        - add `WAIT_FOR_SUBSCRIBER = WAIT_FOR_CONSUMER` for migration purposes
    - rename `QueueFullPolicy::BLOCK_PUBLISHER` -> `QueueFullPolicy::BLOCK_PRODUCER`
        - add `BLOCK_PUBLISHER = BLOCK_PRODUCER` for migration purposes
    - decide if `SubscriberTooSlowPolicy`, `WAIT_FOR_SUBSCRIBER` and `BLOCK_PUBLISHER` should be deprecated

- [ ] 8. Integration test with clients, server and waitsets (user API level)

- [ ] 9. Example

### PR DESCRIPTION
**Pull request by mossmaurice**
## Pre-Review Checklist for the PR Author

1. [x] Code follows the coding style of [CONTRIBUTING.md][contributing]
1. [x] Tests follow the [best practice for testing][testing]
1. [x] ~~Changelog updated [in the unreleased section][changelog] including API breaking changes~~
1. [x] Branch follows the naming format (`iox-#123-this-is-a-branch`)
1. [x] Commits messages are according to this [guideline][commit-guidelines]
    - [x] Commit messages have the issue ID (`iox-#123 commit text`)
    - [x] Commit messages are signed (`git commit -s`)
    - [x] Commit author matches [Eclipse Contributor Agreement][eca] (and ECA is signed)
1. [x] Update the PR title
   - Follow the same conventions as for commit messages
   - Link to the relevant issue
1. [x] Relevant issues are linked
1. [x] Add sensible notes for the reviewer
1. [x] All checks have passed (except `task-list-completed`)
1. [x] Assign PR to reviewer

[commit-guidelines]: https://tbaggery.com/2008/04/19/a-note-about-git-commit-messages.html
[eca]: http://www.eclipse.org/legal/ECA.php
[contributing]: https://github.com/eclipse-iceoryx/iceoryx/blob/master/CONTRIBUTING.md#coding-style
[testing]: https://github.com/eclipse-iceoryx/iceoryx/blob/master/doc/website/advanced/best-practice-for-testing.md
[changelog]: https://github.com/eclipse-iceoryx/iceoryx/blob/master/doc/website/release-notes/iceoryx-unreleased.md

## Notes for Reviewer

* Split up `ServiceDisocvery` tests into `PUB_SUB`-only, `PUB_SUB` && `REQ_RES` and `PUB_SUB` || `REQ_RES`

## Checklist for the PR Reviewer

- [x] Commits are properly organized and messages are according to the guideline
- [x] Code according to our coding style and naming conventions
- [x] Unit tests have been written for new behavior
    - [x] Each unit test case has a unique UUID
- [x] Public API changes are documented via doxygen
- [x] Copyright owner are updated in the changed files
- [x] PR title describes the changes

## Post-review Checklist for the PR Author

1. [x] All open points are addressed and tracked via issues

## References

- Closes #27 (partly)

### CODE REVIEW

File: iceoryx_posh/test/integrationtests/test_service_discovery.cpp [hunk 1]
```diff
@@ -100,348 +123,399 @@ void compareServiceContainers(const ServiceContainer& lhs, const ServiceContaine
...
-    }
+///
+/// Publish-Subscribe & Request-Response
+///
 
-    sut.findService(iox::capro::Wildcard, iox::capro::Wildcard, iox::capro::Wildcard, findHandler);
-    compareServiceContainers(searchResult, searchResultOfFindServiceWithFindHandler);
-}
+using CommunicationKind = Types<Publisher, Server>;
+TYPED_TEST_SUITE(ServiceDiscovery_test, CommunicationKind);
```
Comment (mossmaurice) [inline iceoryx_posh/test/integrationtests/test_service_discovery.cpp:131]: @MatthiasKillat @FerdinandSpitzschnueffler @elBoberido 

We'll get merge conflicts in this file. I don't mind solving these and would propose the following merge order: 

1. #1105 
2. #1127 
3. #1091 

After that I'll tackle #1065.

What do you think?

Comment (elBoberido) [reply iceoryx_posh/test/integrationtests/test_service_discovery.cpp:131]: Sounds reasonable

File: iceoryx_examples/icediscovery/iox_find_service.cpp [hunk 2]
```diff
@@ -48,27 +48,38 @@ int main()
         serviceDiscovery.findService(iox::capro::IdString_t{"Radar"},
                                      iox::capro::IdString_t{"FrontLeft"},
                                      iox::capro::IdString_t{"Image"},
-                                     printSearchResult);
+                                     printSearchResult,
+                                     iox::runtime::MessagingPattern::PUB_SUB);
```
[the remark this answers is not in the document]
Comment (mossmaurice) [reply iceoryx_examples/icediscovery/iox_find_service.cpp:52]: I would prefer to make it explicit to the user and not use an default value. Then there can't be any mix-ups.

File: iceoryx_posh/include/iceoryx_posh/runtime/service_discovery.hpp [hunk 3]
```diff
@@ -48,7 +53,8 @@ class ServiceDiscovery
     /// ServiceContainer: container that is filled with all matching instances
     ServiceContainer findService(const cxx::optional<capro::IdString_t>& service,
                                  const cxx::optional<capro::IdString_t>& instance,
-                                 const cxx::optional<capro::IdString_t>& event) noexcept;
+                                 const cxx::optional<capro::IdString_t>& event,
+                                 const MessagingPattern pattern) noexcept;
```
[the remark this answers is not in the document]
Comment (mossmaurice) [reply iceoryx_posh/include/iceoryx_posh/runtime/service_discovery.hpp:57]: IMHO nope, see [here](https://github.com/eclipse-iceoryx/iceoryx/pull/1134#discussion_r811348927).

File: iceoryx_posh/source/runtime/service_discovery.cpp [hunk 4]
```diff
@@ -23,40 +23,54 @@ namespace runtime
...
     }
 
-    auto lambda = [&](const roudi::ServiceRegistry::ServiceDescriptionEntry& s) {
-        callableForEach(s.serviceDescription);
+    auto lambda = [&](const roudi::ServiceRegistry::ServiceDescriptionEntry& serviceEntry) {
+        switch (pattern)
+        {
+        case MessagingPattern::PUB_SUB:
+        {
+            if (serviceEntry.publisherCount > 0 && serviceEntry.serverCount == 0)
```
[the remark this answers is not in the document]
Comment (mossmaurice) [reply iceoryx_posh/source/runtime/service_discovery.cpp:53]: Good catch! The tests are still failing due to the missing `popo::UntypedClient`.

File: iceoryx_posh/source/runtime/service_discovery.cpp [hunk 5]
```diff
@@ -23,40 +23,54 @@ namespace runtime
...
+        {
+            if (serviceEntry.serverCount > 0 && serviceEntry.publisherCount == 0)
+            {
+                callableForEach(serviceEntry.serviceDescription);
+            }
+            break;
+        }
+        default:
+        {
+            LogWarn() << "ServiceDiscovery could not perform search due to unknown MessagingPattern!";
```
[the remark this answers is not in the document]
Comment (mossmaurice) [reply iceoryx_posh/source/runtime/service_discovery.cpp:69]: We could also reach this branch if someone performs a flawed `static_cast` from a wrong integer.

File: iceoryx_posh/source/runtime/service_discovery.cpp [hunk 6]
```diff
@@ -23,40 +23,54 @@ namespace runtime
...
+                                   const MessagingPattern pattern) noexcept
 {
     if (!callableForEach)
     {
         return;
     }
 
-    auto lambda = [&](const roudi::ServiceRegistry::ServiceDescriptionEntry& s) {
-        callableForEach(s.serviceDescription);
+    auto lambda = [&](const roudi::ServiceRegistry::ServiceDescriptionEntry& serviceEntry) {
```
Comment (MatthiasKillat) [inline iceoryx_posh/source/runtime/service_discovery.cpp:48]: Please create the lambda based on the case, do not resolve the case in the lambda.

Comment (mossmaurice) [reply iceoryx_posh/source/runtime/service_discovery.cpp]: That's more nifty, done.

File: iceoryx_posh/source/runtime/service_discovery.cpp [hunk 7]
```diff
@@ -23,40 +23,54 @@ namespace runtime
...
+        {
+            if (serviceEntry.publisherCount > 0 && serviceEntry.serverCount == 0)
+            {
+                callableForEach(serviceEntry.serviceDescription);
+            }
+            break;
+        }
+        case MessagingPattern::REQ_RES:
+        {
+            if (serviceEntry.serverCount > 0 && serviceEntry.publisherCount == 0)
```
[the remark this answers is not in the document]
Comment (mossmaurice) [reply iceoryx_posh/source/runtime/service_discovery.cpp]: Fixed

File: iceoryx_examples/icediscovery/iox_find_service.cpp (hunk 2 above)
Comment (elBoberido) [inline iceoryx_examples/icediscovery/iox_find_service.cpp:52]: Can we place the `MessagingPattern` into `popo` or `capro`?
It feels weird to be in `runtime`

**Review by elBoberido — COMMENTED** [review]
Looks good. I didn't reviewed the tests since you wrote that there will be merge conflicts. I'd rather wait until they are fixed :)

File: iceoryx_examples/icediscovery/iox_find_service.cpp (hunk 2 above)
Comment (mossmaurice) [reply iceoryx_examples/icediscovery/iox_find_service.cpp:52]: Placed it in `popo` as the `{Publisher,Subscriber}Options` are in there as well. For me it's not part of the user `ServiceDescription` but part of the [internal representation](https://github.com/eclipse-iceoryx/iceoryx/pull/1115#discussion_r811039750).

File: iceoryx_posh/test/integrationtests/test_service_discovery.cpp [hunk 8]
```diff
@@ -540,111 +635,187 @@ TEST_F(ServiceDiscovery_test, NonExistingServicesAreNotFound)
...
 
-TEST_F(ServiceDiscovery_test, ServiceDiscoveryIsNotifiedbyWaitSetAboutSingleService)
+TEST_F(ServiceDiscoveryPubSub_test, ServiceDiscoveryIsNotifiedbyWaitSetAboutSingleService)
 {
     ::testing::Test::RecordProperty("TEST_ID", "f1cf36b5-3db2-4e6f-8e05-e7e449530ec0");
     iox::popo::WaitSet<1U> waitSet;
 
     waitSet
-        .attachEvent(sut,
+        .attachEvent(this->sut,
```
Comment (MatthiasKillat) [inline iceoryx_posh/test/integrationtests/test_service_discovery.cpp:818]: Technically `this` is not needed as it is not a typed test but ok ...

**Review by MatthiasKillat — APPROVED** [review]

File: iceoryx_posh/test/integrationtests/test_service_discovery.cpp (hunk 8 above)
Comment (mossmaurice) [reply iceoryx_posh/test/integrationtests/test_service_discovery.cpp:818]: Please fix this on #1151

**Pull request merged**